src/main/java/com/direwolf20/buildinggadgets/common/building/placement/AbstractTargetIterator.java

```diff
@@ -0,0 +1,146 @@
+package com.direwolf20.buildinggadgets.common.building.placement;
+
+import com.google.common.collect.AbstractIterator;
+import net.minecraft.util.EnumFacing;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.BlockPos.MutableBlockPos;
+
+import javax.annotation.Nonnull;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+/**
+ * Optional predefined iterator logic for implementations of {@link IPlacementSequence}.
+ * <p>
+ * Note that this base class does validation at the same time, which is not recommended to be used in combination of {@link Context}.
+ * In most cases, using the iterator provided by {@link com.direwolf20.buildinggadgets.common.building.Region} is sufficient.
+ * </p>
+ *
+ * @implNote this class reinvented the wheel of {@link AbstractIterator} since it does not expose enough space for further expansion
+ */
+public abstract class AbstractTargetIterator implements Iterator<BlockPos> {
```

Comment:
No need for weird iterator logic when using Streams!

Reply:
To my kownledge, you need a Spliterator to create a stream, and the simplest way to create one is `StreamSupport.stream(iterable.spliterator(), false);` when you have a Iterable<?>, and `Spliterators.spliteratorUnknownSize(iterator, Spliterator.ORDERED), false);` when you only have a Iterator<?>

Reply:
Also `AbstractTargetIterator` is not needed since the validation logic was split out. Will delete in next a few commits

Reply:
You can give it many more characiteristcs, as well as an known size...!
I'll send you an PR which directly implements the Spliterator...

Reply:
You can give it many more characteristcs, expecially you do now the size!
I'll send you a PR which directly implements it, as the [Iterable doc](https://docs.oracle.com/javase/8/docs/api/java/lang/Iterable.html#spliterator--) definitly states, that their implementation is not very efficient.